gdx/src/com/badlogic/gdx/math/Octree.java

```diff
@@ -0,0 +1,307 @@
+
+import com.badlogic.gdx.math.collision.BoundingBox;
+import com.badlogic.gdx.math.collision.Ray;
+import com.badlogic.gdx.utils.Array;
+import com.badlogic.gdx.utils.ObjectSet;
+
+public class Octree<T> {
+
+    private int maxDepth = 8;
+    private int maxItemsPerNode = 32;
+
+    protected OctreeNode root;
+    private final Collider<T> collider;
+
+    public Octree (Vector3 minimum, Vector3 maximum, Collider<T> collider) {
+        super();
+
+        Vector3 realMin = new Vector3(Math.min(minimum.x, maximum.x), Math.min(minimum.y, maximum.y), Math.min(minimum.z, maximum.z));
+        Vector3 realMax = new Vector3(Math.max(minimum.x, maximum.x), Math.max(minimum.y, maximum.y), Math.max(minimum.z, maximum.z));
+
+        this.root = new OctreeNode(realMin, realMax, maxDepth);
+        this.collider = collider;
+    }
+
+    public void add (T aabb) {
+        root.add(aabb, maxItemsPerNode, collider);
+    }
+
+    public void remove (T aabb) {
+        root.remove(aabb);
+    }
+
+    /**
+     * Method to retrieve all the geometries
+     *
+     * @param resultSet
+     * @return the result set
+     */
+    public ObjectSet<T> getAll (ObjectSet<T> resultSet) {
+        root.getAll(resultSet);
+        return resultSet;
+    }
+
+    /**
+     * Method to query geometries inside nodes that the aabb intersects
+     * Can be used as broad phase
+     * @param aabb - The bounding box to query
+     * @param result - Set to be populated with objects inside the BoundingBoxes
+     */
+    public ObjectSet<T> query (BoundingBox aabb, ObjectSet<T> result) {
+        root.query(aabb, result);
+        return result;
+    }
+
+    /**
+     * Method to query geometries inside nodes that the frustum intersects
+     * Can be used as broad phase
+     *
+     * @param frustum - The frustum to query
+     * @param result set populated with objects near from the frustum
+     */
+    public ObjectSet<T> query (Frustum frustum, ObjectSet<T> result) {
+        root.query(frustum, result);
+        return result;
+    }
+
+    public T rayCast (Ray ray) {
+        return rayCast(ray, Float.POSITIVE_INFINITY);
+    }
+
+    public T rayCast (Ray ray, float maxDistance) {
+        RayCastResult<T> result = new RayCastResult<>();
+        result.distance = Float.NEGATIVE_INFINITY;
+        root.rayCast(ray, maxDistance, collider, result);
+        return result.geometry;
+    }
+
+    public void setMaxDepth (int maxDepth) {
+        this.maxDepth = maxDepth;
+        this.root.setLevel(maxDepth);
+    }
+
+    public void setMaxItemsPerNode (int maxItemsPerNode) {
+        this.maxItemsPerNode = maxItemsPerNode;
+    }
+
+    /**
+     * Method to get nodes boxes. Useful for render purpose
+     * @param boxes
+     */
+    public ObjectSet<BoundingBox> getNodesBoxes (ObjectSet<BoundingBox> boxes) {
+        root.getBoundingBox(boxes);
+        return boxes;
+    }
+
+    protected class OctreeNode {
+
+        protected int level;
```

Comment:
should be final

Reply:
I removed the final so it can be used by the node pool. Is it possible to have both?
@mgsx-dev

Reply:
no, we can't have both, this kind of changes are indeed necessary when working with pools.